chown/chown.1
.Dd December 2008
.Dt CHOWN 1

.Sh NAME

.Nm chown
.Nd change the file ownership

.Sh SYNOPSIS

.Nm chown
.Ar [-hR] owner[:group] file ...

.Nm chown
.Ar -R [-H | -L | -P ] owner[:group] file ...

.Sh DESCRIPTION

    The chown utility shall set the user ID of the file named by each file
operand to the user ID specified by the owner operand.

    For each file operand, or, if the -R option is used, each file
encountered while walking the directory trees specified by the file
operands, the chown utility shall perform actions equivalent to the chown()
function defined in the System Interfaces volume of IEEE Std 1003.1-2001,
called with the following arguments:

        The file operand shall be used as the path argument.

        The user ID indicated by the owner portion of the first operand
shall be used as the owner argument.

        If the group portion of the first operand is given, the group ID
indicated by it shall be used as the group argument; otherwise, the group
ownership shall not be changed.

    Unless chown is invoked by a process with appropriate privileges, the
set-user-ID and set-group-ID bits of a regular file shall be cleared upon
successful completion; the set-user-ID and set-group-ID bits of other file
types may be cleared.

.Sh OPTIONS

    The chown utility shall conform to the Base Definitions volume of IEEE
Std 1003.1-2001, Section 12.2, Utility Syntax Guidelines.

    The following options shall be supported by the implementation:

    -h
        If the system supports user IDs for symbolic links, for each file
operand that names a file of type symbolic link, chown shall attempt to set
the user ID of the symbolic link. If the system supports group IDs for
symbolic links, and a group ID was specified, for each file operand that
names a file of type symbolic link, chown shall attempt to set the group ID
of the symbolic link. If the system does not support user or group IDs for
symbolic links, for each file operand that names a file of type symbolic
link, chown shall do nothing more with the current file and shall go on to
any remaining files.
    -H
        If the -R option is specified and a symbolic link referencing a file
of type directory is specified on the command line, chown shall change the
user ID (and group ID, if specified) of the directory referenced by the
symbolic link and all files in the file hierarchy below it.
    -L
        If the -R option is specified and a symbolic link referencing a file
of type directory is specified on the command line or encountered during the
traversal of a file hierarchy, chown shall change the user ID (and group ID,
if specified) of the directory referenced by the symbolic link and all files
in the file hierarchy below it.
    -P
        If the -R option is specified and a symbolic link is specified on
the command line or encountered during the traversal of a file hierarchy,
chown shall change the owner ID (and group ID, if specified) of the symbolic
link if the system supports this operation. The chown utility shall not
follow the symbolic link to any other part of the file hierarchy.
    -R
        Recursively change file user and group IDs. For each file operand
that names a directory, chown shall change the user ID (and group ID, if
specified) of the directory and all files in the file hierarchy below it.
Unless a -H, -L, or -P option is specified, it is unspecified which of these
options will be used as the default.

    Specifying more than one of the mutually-exclusive options -H, -L, and
-P shall not be considered an error. The last option specified shall
determine the behavior of the utility.

.Sh OPERANDS

    The following operands shall be supported:

    owner[:group]
        A user ID and optional group ID to be assigned to file. The owner
portion of this operand shall be a user name from the user database or a
numeric user ID. Either specifies a user ID which shall be given to each
file named by one of the file operands. If a numeric owner operand exists in
the user database as a user name, the user ID number associated with that
user name shall be used as the user ID. Similarly, if the group portion of
this operand is present, it shall be a group name from the group database or
a numeric group ID. Either specifies a group ID which shall be given to each
file. If a numeric group operand exists in the group database as a group
name, the group ID number associated with that group name shall be used as
the group ID.
    file
        A pathname of a file whose user ID is to be modified.

.Sh STDIN

    Not used.

.Sh INPUT FILES

    None.

.Sh ENVIRONMENT VARIABLES

    The following environment variables shall affect the execution of chown:

    LANG
        Provide a default value for the internationalization variables that
are unset or null. (See the Base Definitions volume of IEEE Std 1003.1-2001,
Section 8.2, Internationalization Variables for the precedence of
internationalization variables used to determine the values of locale
categories.)
    LC_ALL
        If set to a non-empty string value, override the values of all the
other internationalization variables.
    LC_CTYPE
        Determine the locale for the interpretation of sequences of bytes of
text data as characters (for example, single-byte as opposed to multi-byte
characters in arguments).
    LC_MESSAGES
        Determine the locale that should be used to affect the format and
contents of diagnostic messages written to standard error.
    NLSPATH
        [XSI] [Option Start] Determine the location of message catalogs for
the processing of LC_MESSAGES . [Option End]

.Sh ASYNCHRONOUS EVENTS

    Default.

.Sh STDOUT

    Not used.

.Sh STDERR

    The standard error shall be used only for diagnostic messages.

.Sh OUTPUT FILES

    None.

.Sh EXTENDED DESCRIPTION

    None.

.Sh EXIT STATUS

    The following exit values shall be returned:

     0
        The utility executed successfully and all requested changes were
made.
    >0
        An error occurred.

.Sh CONSEQUENCES OF ERRORS

    Default.

The following sections are informative.
.Sh APPLICATION USAGE

    Only the owner of a file or the user with appropriate privileges may
change the owner or group of a file.

    Some implementations restrict the use of chown to a user with
appropriate privileges.

.Sh EXAMPLES

    None.

.Sh RATIONALE

    The System V and BSD versions use different exit status codes. Some
implementations used the exit status as a count of the number of errors that
occurred; this practice is unworkable since it can overflow the range of
valid exit status values. These are masked by specifying only 0 and >0 as
exit values.

    The functionality of chown is described substantially through references
to functions in the System Interfaces volume of IEEE Std 1003.1-2001. In
this way, there is no duplication of effort required for describing the
interactions of permissions, multiple groups, and so on.

    The 4.3 BSD method of specifying both owner and group was included in
this volume of IEEE Std 1003.1-2001 because:

        There are cases where the desired end condition could not be
achieved using the chgrp and chown (that only changed the user ID)
utilities. (If the current owner is not a member of the desired group and
the desired owner is not a member of the current group, the chown() function
could fail unless both owner and group are changed at the same time.)

        Even if they could be changed independently, in cases where both are
being changed, there is a 100% performance penalty caused by being forced to
invoke both utilities.

    The BSD syntax user[. group] was changed to user[: group] in this volume
of IEEE Std 1003.1-2001 because the period is a valid character in login
names (as specified by the Base Definitions volume of IEEE Std 1003.1-2001,
login names consist of characters in the portable filename character set).
The colon character was chosen as the replacement for the period character
because it would never be allowed as a character in a user name or group
name on historical implementations.

    The -R option is considered by some observers as an undesirable
departure from the historical UNIX system tools approach; since a tool,
find, already exists to recurse over directories, there seemed to be no good
reason to require other tools to have to duplicate that functionality.
However, the -R option was deemed an important user convenience, is far more
efficient than forking a separate process for each element of the directory
hierarchy, and is in widespread historical use.

.Sh FUTURE DIRECTIONS

    None.

.Sh SEE ALSO

    chmod, chgrp, the System Interfaces volume of IEEE Std 1003.1-2001,
chown()



NEWS

Version 5 - unreleased

	- New utilities: cmp, tput, tabs, stty, uname, uniq

	- Bug fixes.

Version 4 - December 9, 2006

	- import codebase into git repository on git.kernel.org
	- fix compiler warnings on newer compilers
	- add x86-64 support
	- add WIP utility: pax
	- add utilities: env, id, strings, time
	- add simple "it works" tests
	- create per-util status files
	- drop hand-rolled 'compress', import BSD 'compress' instead

Version 3 - February 7, 2005

	- Add copyright/license info to each source file.
	- Make sure bug report address points to SourceForge.

	- New utilities: grep, pathchk, pwd, split, tty.
	- New work-in-progress (WIP) utility: mv.

	- Fix many 'compress' bugs.
	- Security fixes.

	- (internal) Move compress engine into lib.

Version 2 - January 26, 2005

	- Update README with project goals/manifesto,
	  and some URLs.

	- New utilities: compress, comm, date, what, wc, renice,
	  ipcrm, kill, rm.

	- Bug fixes.

Version 1 - August 21, 2004

	- Initial release.



chown/chown.cc

/*
 * Copyright 2004-2006 Jeff Garzik <jgarzik@pobox.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#ifndef HAVE_CONFIG_H
#error missing autoconf-generated config.h.
#endif
#include "posixutils-config.h"

#ifndef _GNU_SOURCE
#define _GNU_SOURCE		/* for O_DIRECTORY */
#endif

#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <string>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <dirent.h>
#include <fcntl.h>
#include <alloca.h>
#include <pwd.h>
#include <grp.h>
#include <libpu.h>

using namespace std;

static const char doc[] =
N_("chown - change file owner and group");

static const char args_doc[] = N_("owner[:group] file ...");

static struct argp_option options[] = {
	{ NULL, 'h', NULL, 0,
	  N_("Set uid/gid of symlink, if file is a symlink.") },
	{ NULL, 'H', NULL, 0,
	  N_("Recurse into symlinked dirs on the command line.") },
	{ NULL, 'L', NULL, 0,
	  N_("Recurse into symlinked dirs on the command line or found during traversal.") },
	{ NULL, 'P', NULL, 0,
	  N_("Do not follow symlinks.") },
	{ NULL, 'R', NULL, 0,
	  N_("Recursively change file user and group IDs.") },
	{ }
};

static error_t parse_opt (int key, char *arg, struct argp_state *state);
static int chown_init(struct walker *w, int argc, char **argv);
static int chown_pre_walk(struct walker *w);
static int chown_arg(struct walker *w, const char *fn, const struct stat *lst);
static int chown_dirent(struct walker *w, const char *dirn, const char *basen,
			const struct stat *lst);

static bool opt_set_link_attr;
static bool opt_set_link_attr_recur;
static bool opt_follow_link_cmdline;
static bool opt_follow_link;
static bool opt_recurse;
static bool opt_chgrp;
static uid_t opt_uid = (uid_t) -1;
static gid_t opt_gid = (gid_t) -1;

static const struct argp argp = { options, parse_opt, args_doc, doc };
static struct walker walker;


static int set_group(const char *group)
{
	unsigned int u;

	struct group *gr = getgrnam(group);
	if (gr == NULL) {
		if (sscanf(group, "%u", &u) != 1) {
			fprintf(stderr, _("invalid group '%s'\n"), group);
			return ARGP_ERR_UNKNOWN;
		}
		opt_gid = u;
	} else
		opt_gid = gr->gr_gid;

	return 0;
}

static int set_owner_group(const char *owner_group)
{
	char *owner, *group;
	int rc, have_group = 0;
	size_t len;
	struct passwd *pw;
	unsigned int u;

	len = strlen(owner_group);
	owner = (char *) alloca(len + 1);
	group = (char *) alloca(len + 1);

	rc = sscanf(owner_group, "%s:%s", owner, group);
	if (rc == 2)
		have_group = 1;
	else if (rc != 1) {
		fprintf(stderr, _("owner[:group] not provided\n"));
		return ARGP_ERR_UNKNOWN;
	}

	pw = getpwnam(owner);
	if (pw == NULL) {
		if (sscanf(owner, "%u", &u) != 1) {
			fprintf(stderr, _("invalid owner '%s'\n"), owner_group);
			return ARGP_ERR_UNKNOWN;
		}
		opt_uid = u;
	} else
		opt_uid = pw->pw_uid;

	if (have_group)
		rc = set_group(group);
	else
		rc = 0;

	return rc;
}

static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
	switch (key) {
	PU_OPT_SET('h', set_link_attr)
	PU_OPT_SET('R', recurse)
	PU_OPT_ARG

	case 'H':
		opt_set_link_attr_recur = false;
		opt_follow_link_cmdline = true;
		opt_follow_link = false;
		break;
	case 'L':
		opt_set_link_attr_recur = false;
		opt_follow_link_cmdline = true;
		opt_follow_link = true;
		break;
	case 'P':
		opt_set_link_attr_recur = true;
		opt_follow_link_cmdline = false;
		opt_follow_link = false;
		break;

	case ARGP_KEY_END:
		if (state->arg_num < 2)		/* not enough args */
			argp_usage (state);
		return ARGP_ERR_UNKNOWN;

	default:
		return ARGP_ERR_UNKNOWN;
	}

	return 0;
}

static void perror_fn(const char *dir, const char *base)
{
	int saved_errno = errno;
	string fn(strpathcat(dir, base));
	errno = saved_errno;

	perror(fn.c_str());
}

static int need_chown (const struct stat *st)
{
	if ((opt_uid != (uid_t)-1) && (st->st_uid != opt_uid))
		return 1;
	if ((opt_gid != (gid_t)-1) && (st->st_gid != opt_gid))
		return 1;
	return 0;
}

static int chown_dirent(struct walker *w, const char *dirn, const char *basen,
			const struct stat *lst)
{
	int is_link = 0, is_dir = 0;

	if (S_ISLNK(lst->st_mode))
		is_link = 1;
	else if (S_ISDIR(lst->st_mode))
		is_dir = 1;

	if (!is_link) {
		/* race between lstat(2) and here */
		if (need_chown(lst) && chown(basen, opt_uid, opt_gid) < 0) {
			perror_fn(dirn, basen);
			w->exit_status = 1;
			return RC_OK;
		}

		if ((!is_dir) || (!opt_recurse))
			return RC_OK;
	}

	else {
		struct stat st;

		if (!opt_follow_link)
			return RC_OK;

		/* race between lstat(2) and here */
		if (stat(basen, &st) < 0) {
			perror_fn(dirn, basen);
			w->exit_status = 1;
			return RC_OK;
		}

		if (!S_ISDIR(st.st_mode))
			return RC_OK;
	}

	return RC_RECURSE;
}

static int chown_arg(struct walker *w, const char *fn, const struct stat *lst)
{
	int is_link = 0, is_dir = 0;

	if (S_ISLNK(lst->st_mode))
		is_link = 1;
	else if (S_ISDIR(lst->st_mode))
		is_dir = 1;

	if (!is_link) {
		/* race between lstat(2) and here */
		if (need_chown(lst) && chown(fn, opt_uid, opt_gid) < 0) {
			perror(fn);
			w->exit_status = 1;
			return RC_OK;
		}

		if ((!is_dir) || (!opt_recurse))
			return RC_OK;
	}

	if (is_link) {
		struct stat st;

		if (opt_set_link_attr) {
			/* race between lstat(2) and here */
			if (lchown(fn, opt_uid, opt_gid) < 0) {
				perror(fn);
				w->exit_status = 1;
			}
			return RC_OK;
		}

		if (!opt_follow_link_cmdline)
			return RC_OK;

		/* race between lstat(2) and here */
		if (stat(fn, &st) < 0) {
			perror(fn);
			w->exit_status = 1;
			return RC_OK;
		}

		if (!S_ISDIR(st.st_mode))
			return RC_OK;
	}

	return RC_RECURSE;
}

static int chown_init(struct walker *w, int argc, char **argv)
{
	pu_init();

	pathelem pe;
	path_split(argv[0], pe);
	if (pe.basen == "chgrp")
		opt_chgrp = true;

	return 0;
}

static int chown_pre_walk(struct walker *w)
{
	string arg = w->arglist.front();
	w->arglist.erase(w->arglist.begin());

	if (opt_chgrp)
		set_group(arg.c_str());
	else
		set_owner_group(arg.c_str());

	if (!opt_recurse) {
		opt_set_link_attr_recur = false;
		opt_follow_link_cmdline = false;
		opt_follow_link = false;
	} else {
		w->flags |= WF_RECURSE;
		if (opt_set_link_attr_recur)
			opt_set_link_attr = true;
	}

	if (opt_follow_link)
		w->flags |= WF_FOLLOW_LINK;
	if (opt_follow_link_cmdline)
		w->flags |= WF_FOLLOW_LINK_CMDLINE;

	return 0;
}

int main (int argc, char *argv[])
{
	walker.argp			= &argp;
	walker.flags			= WF_STAT;
	walker.init			= chown_init;
	walker.pre_walk			= chown_pre_walk;
	walker.cmdline_arg		= chown_arg;
	walker.dirent			= chown_dirent;
	return walk(&walker, argc, argv);
}
